pgjdbc/src/main/java/org/postgresql/jdbc/ArrayDecoding.java

```diff
@@ -0,0 +1,794 @@
+      return bytes.getFloat();
+    }
+
+    @Override
+    Object parseValue(String stringVal, BaseConnection connection) throws SQLException {
+      return PgResultSet.toFloat(stringVal);
+    }
+  };
+
+  private static final ArrayDecoder<Boolean[]> BOOLEAN_OBJ_ARRAY = new AbstractObjectArrayDecoder<Boolean[]>(
+      Boolean.class) {
+
+    @Override
+    Object parseValue(int length, ByteBuffer bytes, BaseConnection connection) {
+      return bytes.get() == 1;
+    }
+
+    @Override
+    Object parseValue(String stringVal, BaseConnection connection) throws SQLException {
+      return BooleanTypeUtil.fromString(stringVal);
+    }
+  };
+
+  private static final ArrayDecoder<String[]> STRING_ARRAY = new AbstractObjectArrayDecoder<String[]>(String.class) {
+
+    @Override
+    Object parseValue(int length, ByteBuffer bytes, BaseConnection connection) throws SQLException {
+      assert bytes.hasArray();
+      final byte[] byteArray = bytes.array();
+      final int offset = bytes.arrayOffset() + bytes.position();
+
+      String val;
+      try {
+        val = connection.getEncoding().decode(byteArray, offset, length);
+      } catch (IOException e) {
+        throw new PSQLException(GT.tr(
+            "Invalid character data was found.  This is most likely caused by stored data containing characters that are invalid for the character set the database was created in.  The most common example of this is storing 8bit data in a SQL_ASCII database."),
+            PSQLState.DATA_ERROR, e);
+      }
+      bytes.position(bytes.position() + length);
+      return val;
+    }
+
+    @Override
+    Object parseValue(String stringVal, BaseConnection connection) throws SQLException {
+      return stringVal;
+    }
+  };
+
+  private static final ArrayDecoder<byte[][]> BYTE_ARRAY_ARRAY = new AbstractObjectArrayDecoder<byte[][]>(
+      byte[].class) {
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    Object parseValue(int length, ByteBuffer bytes, BaseConnection connection) throws SQLException {
+      final byte[] array = new byte[length];
+      bytes.get(array);
+      return array;
+    }
+
+    @Override
+    Object parseValue(String stringVal, BaseConnection connection) throws SQLException {
+      try {
+        return PGbytea.toBytes(stringVal.getBytes("ascii"));
+      } catch (UnsupportedEncodingException e) {
+        throw new java.lang.Error("ascii must be supported");
+      }
+    }
+  };
+
+  private static final ArrayDecoder<BigDecimal[]> BIG_DECIMAL_DECODER = new AbstractObjectArrayDecoder<BigDecimal[]>(
+      BigDecimal.class) {
+
+    @Override
+    Object parseValue(String stringVal, BaseConnection connection) throws SQLException {
+      return PgResultSet.toBigDecimal(stringVal);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    Object parseValue(int length, ByteBuffer bytes, BaseConnection connection) throws SQLException {
```

Comment:
This has support for binary, but currently the PGConnection does not treat Oid.NUMERIC_ARRAY (or even Oid.NUMERIC) as binary supported.
My preference would be to remove this and continue to leave NUMERIC types only supported with String.

Reply:
Is there some reason we can't support NUMERIC in binary form ?

Reply:
I should have expanded.
I do not believe we should add binary support for NUMERIC as part of this pr.
I have no issues with binary support conceptually, but it should be it’s own pr which covers all uses (arrays are pretty small part, really), including unit tests.

Reply:
Seems reasonable.